pennylane_lightning/lightning_tensor/_tensornet.py

```diff
@@ -31,44 +31,100 @@
-    U, S, Vd = np.linalg.svd(M, full_matrices=False)
+def svd_split(Mat, site_shape, max_bond_dim):
+    """SVD decomposition of a matrix via numpy linalg. Note that this function is to be moved to the C++ layer."""
+    # TODO: Check if cutensornet allows us to remove all zero (or < tol) singular values and the respective rows and columns of U and Vd
+    U, S, Vd = np.linalg.svd(Mat, full_matrices=False)
     U = U @ np.diag(S)  # Append singular values to U
     bonds = len(S)
-    Vd = Vd.reshape(bonds, 2, -1)
-    U = U.reshape((-1, 2, bonds))
+
+    Vd = Vd.reshape(tuple([bonds] + site_shape + [-1]))
+    U = U.reshape(tuple([-1] + site_shape + [bonds]))
 
     # keep only chi bonds
-    chi = np.min([bonds, bond_dim])
-    U, S, Vd = U[:, :, :chi], S[:chi], Vd[:chi]
+    chi = np.min([bonds, max_bond_dim])
+    U, Vd = U[..., :chi], Vd[:chi]
     return U, Vd
 
 
-def dense_to_mps(psi, n_wires, bond_dim):
-    """Convert a dense state vector to a matrix product state."""
+def decompose_dense(psi, n_wires, site_shape, max_bond_dim):
+    """Decompose a dense state vector/gate matrix into MPS/MPO sites."""
     Ms = [[] for _ in range(n_wires)]
+    site_len = np.prod(site_shape)
+    psi = np.reshape(psi, (site_len, -1))  # split psi [2, 2, 2, 2...] to psi [site_len, -1]
 
-    psi = np.reshape(psi, (2, -1))  # split psi[2, 2, 2, 2..] = psi[2, (2x2x2...)]
-    U, Vd = svd_split(psi, bond_dim)  # psi[2, (2x2x..)] = U[2, mu] Vd[mu, (2x2x2x..)]
+    U, Vd = svd_split(
+        psi, site_shape, max_bond_dim
+    )  # psi [site_len, -1] -> U [site_len, mu] Vd [mu, (2x2x2x..)]
 
-    Ms[0] = U
+    Ms[0] = U.reshape(tuple(site_shape + [-1]))
     bondL = Vd.shape[0]
     psi = Vd
 
     for i in range(1, n_wires - 1):
-        psi = np.reshape(psi, (2 * bondL, -1))  # reshape psi[2 * bondL, (2x2x2...)]
-        U, Vd = svd_split(psi, bond_dim)  # psi[2, (2x2x..)] = U[2, mu] Vd[mu, (2x2x2x..)]
+        psi = np.reshape(psi, (site_len * bondL, -1))  # reshape psi[site_len*bondL, -1]
+        U, Vd = svd_split(
+            psi, site_shape, max_bond_dim
+        )  # psi [site_len*bondL, -1] -> U [site_len, mu] Vd [mu, (2x2x2x..)]
         Ms[i] = U
 
         psi = Vd
         bondL = Vd.shape[0]
 
-    Ms[n_wires - 1] = Vd
+    Ms[-1] = Vd.reshape(tuple([-1] + site_shape))
 
     return Ms
 
 
+def gate_matrix_decompose(gate_ops_matrix, wires, c_dtype):
+    """Permute and decompose a gate matrix into MPO sites. This method return the MPO sites in the Fortran order of the ``cutensornet`` backend. Note that MSB in the Pennylane convention is the LSB in the ``cutensornet`` convention."""
+    sorted_indexed_wires = sorted(enumerate(wires), key=lambda x: x[1])
+
+    sorted_wires = []
+    original_axes = []
+    for index, wire in sorted_indexed_wires:
+        sorted_wires.append(wire)
+        original_axes.append(index)
```

Comment:
Hmmm.. maybe better to use zip to avoid multiple append and list resize:
```suggestion
    sorted_wires, original_axes = zip(*sorted_indexed_wires)
```
We also don't need to convert them to list

Reply:
Sure!